docs/ProofOptimization/ProofOptimization.md

```diff
@@ -0,0 +1,599 @@
+
+<!-- %no-check -->
+```text
+...
+  Assertion batch 4:
+    Outcome: Valid
+    Duration: 00:00:00.1344230
+    Resource count: 63182
+
+    Assertions:
+      ProveSomeArithmetic.dfy(6,22): result of operation never violates newtype constraints for 'uint32'
+
+  Assertion batch 5:
+    Outcome: Valid
+    Duration: 00:00:00.1359990
+    Resource count: 169237
+
+    Assertions:
+      ProveSomeArithmetic.dfy(6,32): assertion always holds
+
+  Assertion batch 6:
+    Outcome: Valid
+    Duration: 00:00:00.0844010
+    Resource count: 63494
+
+    Assertions:
+      ProveSomeArithmetic.dfy(7,14): result of operation never violates newtype constraints for 'uint32'
+...
+```
+
+The `--log-format csv` option to `dafny verify`, or the `/verificationLogger:csv` old-style option, produces output in a file in CSV format, and prints the name of the file. For the above method, this file might look something like the following.
+
+<!-- %no-check -->
+```text
+TestResult.DisplayName,TestResult.Outcome,TestResult.Duration,TestResult.ResourceCount
+ProveSomeArithmetic (correctness) (assertion batch 1),Passed,00:00:00.1880600,50027
+ProveSomeArithmetic (correctness) (assertion batch 2),Passed,00:00:00.1640300,127152
+ProveSomeArithmetic (correctness) (assertion batch 3),Passed,00:00:00.1389720,59361
+ProveSomeArithmetic (correctness) (assertion batch 4),Passed,00:00:00.1296620,63182
+ProveSomeArithmetic (correctness) (assertion batch 5),Passed,00:00:00.1232840,169237
+ProveSomeArithmetic (correctness) (assertion batch 6),Passed,00:00:00.0937380,63494
+ProveSomeArithmetic (correctness) (assertion batch 10),Passed,00:00:00.0774500,66869
+ProveSomeArithmetic (correctness) (assertion batch 9),Passed,00:00:00.0708400,67341
+ProveSomeArithmetic (correctness) (assertion batch 8),Passed,00:00:00.0644100,64792
+ProveSomeArithmetic (correctness) (assertion batch 7),Passed,00:00:00.0622940,65952
+```
+
+In all of these output formats, you can see that batch 5, corresponding to the assertion about conversion to and from bit vectors, is the most expensive to prove. Later sections will go into why this is the case, and how to make proofs involving bit vectors more straightforward.
+
+# Identifying highly variable assertions
+
+Sometimes, verifying a particular assertion is fast, initially, but becomes very slow (or fails) after a small change. We refer to this phenomenon as _verification variability_. The techniques in this document can help to fix proofs when it occurs. However, sometimes it would be more helpful to predict when verification of your program is highly variable, and adapt it proactively, when you can intentionally set aside time, rather than needing to fix it retroactively.
```

Comment:
```suggestion
Sometimes, verifying a particular assertion succeeds, initially, but fails after a small change. We refer to this phenomenon as _verification variability_. The techniques in this document can help to fix proofs when it occurs. However, sometimes it would be more helpful to predict when verification of your program is likely to be highly variable, and adapt it proactively, when you can intentionally set aside time, rather than needing to fix it retroactively.
```

Often variable "proofs" are already slow to verify.